src/sequence/mod.rs

```diff
@@ -719,22 +724,368 @@ mod tests {
+
+            let sdata_tag = 43_001u64;
+
+            // Instantiate the same Sequence on two replicas
+            let mut replica1 = Sequence::new_public(actor1, sequence_name, sdata_tag);
+            let mut replica2 = Sequence::new_public(actor2, sequence_name, sdata_tag);
+
+            // Set Actor1 as the owner
+            let perms = BTreeMap::default();
+            let owner_op = replica1.set_public_policy(actor1, perms)?;
+            replica2.apply_public_policy_op(owner_op)?;
+
+            let dataset_length = dataset.len() as u64;
+
+            let mut ops = vec![];
+            for data in dataset {
+                    let op = replica1.append(data)?;
+                    ops.push(op);
+            }
+
+            for op in ops.clone() {
+                let delivery_chance : u32 = OsRng.gen_range(0, 10);
+                if delivery_chance < 2 {
+                    replica2.apply_data_op(op)?;
+                }
+            }
+
+            // shere we statistically should have dropped some messages
+            assert_ne!(replica2.len(), replica1.len());
+
+            // reapply all ops
+            for op in ops {
+                    replica2.apply_data_op(op)?;
+            }
+
+            // now we converge
+            verify_data_convergence(vec![replica1, replica2], dataset_length);
+
+        }
+
+
+        #[test]
+        fn proptest_converge_with_shuffled_ops_from_many_while_dropping_some_at_random(
+            dataset in generate_dataset(1000),
+            mut replicas in generate_replicas(100),
+        ) {
+            let dataset_length = dataset.len() as u64;
+
+            // generate an ops set using random replica for each data
+            let mut ops = vec![];
+            for data in dataset {
+                if let Some(replica) = replicas.choose_mut(&mut OsRng)
+                {
+                    let op = replica.append(data)?;
+                    ops.push(op);
+                }
+            }
+
+            let opslen = ops.len() as u64;
+            prop_assert_eq!(dataset_length, opslen);
+
+            // now we randomly shuffle ops and apply at each replica
+            for replica in &mut replicas {
+
+                let mut ops = ops.clone();
+                ops.shuffle(&mut OsRng);
+
+                for op in ops {
+                    let delivery_chance : u32 = OsRng.gen_range(0, 10);
+                    if delivery_chance > 3 {
+                        replica.apply_data_op(op)?;
+                    }
+                }
+            }
+
+            // TODO: Currently this wont converge. Data is missing.
+            // We need to be attempting to apply ops which are known to be not causally ready...
+            // We can get errors there by dropping from sync with one replica...
+            //
+            // Q: how to expand this across many replicas and simulate dropping messages? Is that wortwhile here at the data itself...?
```

Comment:
It seems we don't need to test that here as that would be part of upper layers logic, like the Elders syncing up from time to time?
I think here at this layer we shod just ensure that if all ops were applied in any order then the data converges, and as you did here perhaps also test that i any op is missing then they don't converge, etc.

Reply:
> It seems we don't need to test that here as that would be part of upper layer's logic, like the Elders syncing up from time to time?

Cools. I was thinking similar but wasn't 100% sure 👍